phac-nml/irida-next
# Attachments: Illumina Paired End detection

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR updates the `Attachments::CreateService` to identify Illumina Paired End attachments when more than 1 `fastq` file is successfully attached to the attachable.

When both the `forward` and `reverse` file is detected for `fastq` files that match the [Illumina Fastq naming convention](https://help.basespace.illumina.com/files-used-by-basespace/fastq-files#naming) add the following metadata:

Forward:
```json
{
  "type": "illumina_pe",
  "direction": "forward",
  "associated_attachment_id": 2
}
```

Reverse:
```json
{
  "type": "illumina_pe",
  "direction": "reverse",
  "associated_attachment_id": 1
}
```

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Login as your user of choice, and create a personal project
2. Create a Sample in the project you just created
3. Navigate to the Sample
4. Click `Upload Files`
5. Select `test/fixtures/files/TestSample_S1_L001_R1_001.fastq` and `test/fixtures/files/TestSample_S1_L001_R2_001.fastq`
6. Click `Upload`
7. In your Rails console, execute `Attachments.last` see that it has the following metadata fields `type`, `direction`, and `associated_attachment_id`
8. In your Rails console, execute `Attachments.last.associated_attachment` see that it has the following metadata fields `type`, `direction`, and `associated_attachment_id`

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
